### PR TITLE
rocAL - Fix issues with cascaded augmentations

### PR DESCRIPTION
**Pull request by swetha097**
Fix issues with cascaded augmentations for gpu backend

### CODE REVIEW

File: rocAL/rocAL/rocAL_hip/rali_hip_kernels.cpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ Hip_CopyInt8ToNHWC_fp32
         if (C == 3){
                 float3 dst;
```
Comment (rrawther) [inline rocAL/rocAL/rocAL_hip/rali_hip_kernels.cpp:52]: please adjust spacing here to be consistant

File: rocAL/rocAL/source/master_graph.cpp [hunk 2]
```diff
@@ -659,12 +659,12 @@ MasterGraph::copy_out_tensor(void *out_ptr, RaliTensorFormat format, float multi
             auto img_buffer = out_image;
             if (format == RaliTensorFormat::NHWC)
             {
-                HipExecCopyInt8ToNHWC(_device.resources().hip_stream, (const void *)img_buffer, out_ptr, dest_buf_offset, n, c, h, w,
+                HipExecCopyInt8ToNHWC(_device.resources().hip_stream, (const void *)img_buffer, out_ptr, dest_buf_offset, n , c, h, w,
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/source/master_graph.cpp:662]: extra space here

File: rocAL/rocAL/source/master_graph.cpp [hunk 3]
```diff
@@ -659,12 +659,12 @@ MasterGraph::copy_out_tensor(void *out_ptr, RaliTensorFormat format, float multi
...
                                         multiplier0, multiplier1, multiplier2, offset0, offset1, offset2, reverse_channels, fp16);
 
             }else
             {
-                HipExecCopyInt8ToNCHW(_device.resources().hip_stream, (const void *)img_buffer, out_ptr, dest_buf_offset, n, c, h, w,
+                HipExecCopyInt8ToNCHW(_device.resources().hip_stream, (const void *)img_buffer, out_ptr, dest_buf_offset, n , c, h, w,
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/source/master_graph.cpp:667]: extra space here

File: rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py [hunk 4]
```diff
@@ -43,7 +46,7 @@ def main():
     except OSError as error:
         print(error)
     # Create Pipeline instance
-    pipe = Pipeline(batch_size=batch_size, num_threads=num_threads, device_id=args.local_rank, seed=random_seed, rali_cpu=rali_cpu)
+    pipe = Pipeline(batch_size=batch_size, num_threads=num_threads, device_id=args.local_rank, seed=random_seed, rali_cpu=rali_cpu, tensor_layout=types.NHWC if args.NHWC else args.NCHW , tensor_dtype=types.FLOAT16 if args.fp16 else types.FLOAT)
```
Comment (LakshmiKumar23) [inline rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py:49]: should be 
```
tensor_layout=types.NHWC if args.NHWC else types.NCHW
```
Please check

File: rocAL/rocAL/source/master_graph.cpp [hunk 5]
```diff
@@ -659,12 +659,12 @@ MasterGraph::copy_out_tensor(void *out_ptr, RaliTensorFormat format, float multi
             auto img_buffer = out_image;
             if (format == RaliTensorFormat::NHWC)
             {
```
Comment (rrawther) [inline rocAL/rocAL/source/master_graph.cpp:661]: please revert the changes in this file since nothing has changed

File: rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py [hunk 6]
```diff
@@ -92,7 +94,7 @@ def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = Fals
                     self.out = torch.empty((self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype=torch.float32)
                 elif self.tensor_dtype == types.FLOAT16:
                     self.out = torch.empty((self.bs*self.n, int(self.h/self.bs), self.w, self.p), dtype=torch.float16)
-                self.labels = torch.empty(self.labels_size, dtype = torch.int32, device = torch_gpu_device)
+                self.labels = torch.empty(self.labels_size, dtype = torch.int32)
```
Comment (rrawther) [inline rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:97]: why is this changed back to cpu allocation. I thought only TF needed this change

File: rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh [hunk 7]
```diff
@@ -97,36 +97,36 @@ rocAL_api_python_unittest=1
...
-    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name resize_brightness_jitter --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 --$backend_arg 2>&1 | tee -a run.rocAL_api_log.${CURRENTDATE}
-    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name vignetter_blur --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 --$backend_arg 2>&1 | tee -a run.rocAL_api_log.${CURRENTDATE}
-    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name snow --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 2 --$backend_arg 2>&1 | tee -a run.rocAL_api_log.${CURRENTDATE}
+    python$ver rocAL_api_python_unittest.py --image-dataset-path $data_dir --augmentation-name resize --batch-size $batch_size --display --NHWC --local-rank 0 --world-size $gpus_per_node --num-threads 1 --num-epochs 1 --$backend_arg 2>&1 | tee -a run.rocAL_api_log.${CURRENTDATE}
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh:100]: I think, it is better to keep it as 2 epochs so we can get more test coverage. Is there any reason why it is needed?

File: rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py [hunk 8]
```diff
@@ -16,8 +16,11 @@ def draw_patches(img, idx, device):
             image = img.detach().numpy()
     else:
         image = img.cpu().numpy()
-    image = image.transpose([1, 2, 0])
     image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
+    if not args.NHWC:
+        image = image.transpose([1, 2, 0])
+    if args.fp16:
+        image = (image).astype('uint8')
```
Comment (rrawther) [inline rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py:23]: Is this correct. For fp16, we are converting to uint8

**Review by rrawther — CHANGES_REQUESTED** [review]
@swetha097 : please address requested changes

File: rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py (hunk 8 above)
Comment (LakshmiKumar23) [reply rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py:23]: I think @rrawther  that's for display, I was going to comment too

File: rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh (hunk 7 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/new_api/PYTHON_UNITTEST_TEST_FILE.sh:100]: No specific reason Rajy.

I will change it back to 2 epochs.

File: rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py (hunk 6 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py:97]: This was a bug Rajy.

I had used gpu device for a cpu tensor allocation.

We do have the gpu allocation also.

File: rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py (hunk 8 above)
Comment (swetha097) [reply rocAL/rocAL_pybind/example/new_api/rocAL_api_python_unittest.py:23]: Its for opencv imwrite.

we need to convert the fp16 to uint8 for dumping images.

File: rocAL/rocAL/source/master_graph.cpp (hunk 2 above)
Comment (swetha097) [reply rocAL/rocAL/source/master_graph.cpp]: Reverted the changes in master graph

File: rocAL/rocAL/source/master_graph.cpp (hunk 3 above)
Comment (swetha097) [reply rocAL/rocAL/source/master_graph.cpp]: reverted the changes in master graph

File: rocAL/rocAL/source/master_graph.cpp (hunk 5 above)
Comment (swetha097) [reply rocAL/rocAL/source/master_graph.cpp]: reverted

**Comment by LakshmiKumar23** [comment]
@swetha097 Please turn off the tensor dump by default. 
Fixes the cascade issue

File: rocAL/rocAL/source/master_graph.cpp [hunk 9]
```diff
@@ -659,12 +659,12 @@ MasterGraph::copy_out_tensor(void *out_ptr, RaliTensorFormat format, float multi
             auto img_buffer = out_image;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/master_graph.cpp:659]: Have already addressed this Rajy

[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/source/master_graph.cpp:659]: The files changed now doesn't show master_graph.cpp

**Comment by kiritigowda** [comment]
Issue #813 - Fix

**Review by rrawther — APPROVED** [review]
@LakshmiKumar23 to verify and test

**Comment by kiritigowda** [comment]
@LakshmiKumar23  is this good to merge?

**Comment by LakshmiKumar23** [comment]
@kiritigowda Yes, this can be merged

**Review by LakshmiKumar23 — APPROVED** [review]
